### PR TITLE
Add cart functionality

### PR DESCRIPTION
**Pull request by dylanerichards**
# Summary

This PR adds the concept of a cart, and the ability for a user to add items to their shopping cart.

#Trello Card 

https://trello.com/c/2wvqmfkG/16-cart-feature-ability-to-checkout

# Follow ups

- Removing items from the cart will be added in a subsequent PR

### CODE REVIEW

File: .byebug_history [hunk 1]
```diff
@@ -0,0 +1,34 @@
+continue
```
Comment (dylanerichards) [inline .byebug_history]: This is why you should never do `git add .`. You never know what you're committing blindly. This should be added to the `.gitignore`.

File: Gemfile [hunk 2]
```diff
@@ -13,6 +13,7 @@ gem "stimulus-rails"
 gem "jbuilder"
 gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]
 gem "bootsnap", require: false
+gem "byebug"
```
Comment (dylanerichards) [inline Gemfile:16]: A good gem to have around for debugging.

File: app/controllers/cart_products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class CartProductsController < ApplicationController
+  def create
```
Comment (dylanerichards) [inline app/controllers/cart_products_controller.rb:2]: This controller action is where the business logic for adding an item to a cart currently lives. Eventually, we will refactor this logic to a service object or to the `Cart` model.

File: app/controllers/cart_products_controller.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    products_in_cart = current_user.cart.products
+
+    if products_in_cart.map(&:id).include?(params[:product_id].to_i)
```
Comment (dylanerichards) [inline app/controllers/cart_products_controller.rb:5]: Here, we check to see if the products that currently exist in the cart already include the item we're trying to add to our cart.

File: app/controllers/cart_products_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      product_in_cart = products_in_cart.where(id: params[:product_id]).first
```
Comment (dylanerichards) [inline app/controllers/cart_products_controller.rb:6]: If the product exists in our cart, we want to locate that product.

File: app/controllers/cart_products_controller.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+      cart_product = CartProduct.find_by(product_id: product_in_cart, cart_id: current_user.cart.id)
```
Comment (dylanerichards) [inline app/controllers/cart_products_controller.rb:8]: Once we have the product, we can look up the join model, `CartProduct` and update its quantity.

File: app/controllers/cart_products_controller.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      cart_product.update(quantity: cart_product.quantity + 1)
+      cart_product.save
+    else
```
Comment (dylanerichards) [inline app/controllers/cart_products_controller.rb:11]: If the product does not already exist in the cart then we just want to create a new `CartProduct` record.

**Comment by rjackson21** [comment]
Hey Dylan - thanks for pairing with me yesterday. My biggest takeaways from this are join table we created (this was my first time examining what they do) and of course debugging with byebug. Byebug was essential here to help us with the if statement in the `cart_products_controller`, and I assume it will continue to be a great help for future projects. All of the changes look good to me! I'm looking forward to styling the cart page next.

**Review by rjackson21 — APPROVED** [review]

**Pull request merged**